Rewrite TINC test for freezing auxiliary tables on an empty relation.

hlinnaka opened the pull request:
These tests test for an old bug, where vacuuming an empty AO relation would
not advance its relfrozenxid. Doesn't seem very likely for that particular
bug to reappear, but seems like a good idea to have some coverage for
freezing and relfrozenxid advancement. In this rewritten form, these tests
are fairly quick, too (1-2 seconds on my laptop).

ashwinstar commented:
Can you please clarify in the PR what's the difference between old and re-written form to easily review and understand the difference ? Does the coverage remain exactly the same or reduces ?

hlinnaka commented:
On 02/21/2017 10:01 AM, Ashwin Agrawal wrote:
> Can you please clarify in the PR what's the difference between old
> and re-written form to easily review and understand the difference ?
> Does the coverage remain exactly the same or reduces ?

The basic flow of the tests is the same:

1. CREATE TABLE, leave it empty (or in some variations, insert some 
data, then DELETE all the data to make it empty again)
2. Run some commands, to advance XID counter
3. VACUUM
4. Verify that the relfrozenxid was advanced by the vacuum

Now, the differences:

* Instead of having a separate test file for heap, AO, and AOCO tables, 
they are all in the same file. It first creates all three tables, then 
advances the XID counter, then VACUUMs all three tables, then verifies 
all three tables.

* The old tests tested an AO and an AOCO table that had a toast table, 
and another that did not. That seemed excessive, so I left out the tests 
without a toast table. If it works with a toast table, hard to imagine 
it failing without one.

* The new tests also test a heap table, while the old tests only tested 
an AO and AOCO table. Maybe not really necessary, but seems nice for 
completeness.

* The old tests launched psql to run a dummy SELECT query 150 times, to 
advance the XID counter. The new test runs a PL/pgSQL function with a 
loop over an exception block instead (an exception block creates a 
subtransaction, which gets assigned an XID). That's a lot faster. And 
instead of a SELECT in the loop, the new test uses an INSERT, so that 
the test still works if we get lazy-XID-assignment working on GPDB.

* Verifying the age of relfrozenxid uses a little PL/pgSQL helper 
function, instead of running queries in the Python harness. Similar 
effect, but different implementation.

So, the coverage isn't exactly the same, but I believe it covers well 
what this test was meant to cover. I believe these tests were added in 
response to bug MPP-24168 ("Vacuum did not update relfrozenxid for UAO 
underlying tables if the table contains NO data"), and that is still 
covered.

- Heikki

ashwinstar commented:
On Tue, Feb 21, 2017 at 12:23 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> On 02/21/2017 10:01 AM, Ashwin Agrawal wrote:
> > Can you please clarify in the PR what's the difference between old
> > and re-written form to easily review and understand the difference ?
> > Does the coverage remain exactly the same or reduces ?
>
> The basic flow of the tests is the same:
>
> 1. CREATE TABLE, leave it empty (or in some variations, insert some
> data, then DELETE all the data to make it empty again)
> 2. Run some commands, to advance XID counter
> 3. VACUUM
> 4. Verify that the relfrozenxid was advanced by the vacuum
>

Looks great comment to have at top of test SQL file.


> Now, the differences:
>
> * Instead of having a separate test file for heap, AO, and AOCO tables,
> they are all in the same file. It first creates all three tables, then
> advances the XID counter, then VACUUMs all three tables, then verifies
> all three tables.
>
> * The old tests tested an AO and an AOCO table that had a toast table,
> and another that did not. That seemed excessive, so I left out the tests
> without a toast table. If it works with a toast table, hard to imagine
> it failing without one.
>

Feel should have same or more coverage as past, hence should have case
without toast as well. Atleast for one of AO/AOCO table if not all three
combinations of tables.


>
> * The new tests also test a heap table, while the old tests only tested
> an AO and AOCO table. Maybe not really necessary, but seems nice for
> completeness.
>

Nice, thumbs up for added coverage.


> * The old tests launched psql to run a dummy SELECT query 150 times, to
> advance the XID counter. The new test runs a PL/pgSQL function with a
> loop over an exception block instead (an exception block creates a
> subtransaction, which gets assigned an XID). That's a lot faster. And
> instead of a SELECT in the loop, the new test uses an INSERT, so that
> the test still works if we get lazy-XID-assignment working on GPDB.
>

Wonderfull.

hlinnaka commented:
On 02/21/2017 10:42 AM, Ashwin Agrawal wrote:
> On Tue, Feb 21, 2017 at 12:23 AM, Heikki Linnakangas <
> notifications@github.com> wrote:
>
>> On 02/21/2017 10:01 AM, Ashwin Agrawal wrote:
>>> Can you please clarify in the PR what's the difference between old
>>> and re-written form to easily review and understand the difference ?
>>> Does the coverage remain exactly the same or reduces ?
>>
>> The basic flow of the tests is the same:
>>
>> 1. CREATE TABLE, leave it empty (or in some variations, insert some
>> data, then DELETE all the data to make it empty again)
>> 2. Run some commands, to advance XID counter
>> 3. VACUUM
>> 4. Verify that the relfrozenxid was advanced by the vacuum
>
> Looks great comment to have at top of test SQL file.

True, added.

>> * The old tests tested an AO and an AOCO table that had a toast table,
>> and another that did not. That seemed excessive, so I left out the tests
>> without a toast table. If it works with a toast table, hard to imagine
>> it failing without one.
>
> Feel should have same or more coverage as past, hence should have case
> without toast as well. Atleast for one of AO/AOCO table if not all three
> combinations of tables.

It's easy to go crazy with different combinations, without adding any 
meaningful coverage. But I added the no-toast variants now, like in the 
old tests.

- Heikki

ashwinstar commented on src/test/regress/atmsort.pm:
```diff
@@ -1113,7 +1113,7 @@ sub atmsort_bigloop
     my $define_match_expression = undef;
 
-    print $atmsort_outfh "GP_IGNORE: formatted by atmsort.pm\n";
+#    print $atmsort_outfh "GP_IGNORE: formatted by atmsort.pm\n";
```
Delete the line, instead of commenting  ?

hlinnaka commented:
On 02/21/2017 11:19 AM, Ashwin Agrawal wrote:
> ashwinstar approved this pull request.
>
>
>
>> @@ -1113,7 +1113,7 @@ sub atmsort_bigloop
>      my $big_ignore = 0;
>      my $define_match_expression = undef;
>
> -    print $atmsort_outfh "GP_IGNORE: formatted by atmsort.pm\n";
> +#    print $atmsort_outfh "GP_IGNORE: formatted by atmsort.pm\n";
>
> Delete the line, instead of commenting  ?

D'oh, I wasn't supposed to commit that. I will leave out that change 
before pushing.

This probably needs an explanation though:

The "formatted by atmsort.pm" line gets in the way, when I open 
regression.diffs in emacs, and try to apply the changes to expected 
output file, one hunk at a time. If there are any change at the top of a 
file, like a comment explaining the purpose of the test file, that extra 
"formatted by atmsort.pm" line that gpdiff.pm prints, prevents the diff 
from applying. So I tend to temporarily comment that out.

Perhaps we should remove that from atmsort.pm permanently. It might be 
useful to be able to tell from the .out file, if it's been processed by 
atmsort.pm, in some scenarios. Although I don't recall encountering one. 
For me at least, it just gets in the way.

- Heikki

hlinnaka commented:
Pushed, thanks for the review!
